# Perform feature engineering 

## **Introduction**

In this activity, I’m stepping into the role of a data analyst for a firm that delivers insights to the National Basketball Association (NBA)—one of the most exciting and competitive leagues in the world. My goal is to help NBA managers and coaches uncover which players are most likely to succeed in the intense, high-stakes environment of professional basketball and contribute to their teams' long-term success. 

To make that happen, I’ll be diving into a rich dataset of NBA player performance stats. Through thoughtful feature engineering, I’ll identify the most meaningful variables that can help predict whether a player’s career in the NBA will last at least five years. The findings from this analysis will lay the groundwork for the next phase of the project: building a powerful predictive model.


## ** Imports** 


In [1]:
# Import pandas.

import pandas as pd

The dataset is a .csv file named `nba-players.csv`. It consists of performance records for a subset of NBA players. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# RUN THIS CELL TO IMPORT YOUR DATA.
# Save in a variable named `data`.
data = pd.read_csv("nba-players.csv", index_col=0)

## **Data exploration** 

In [3]:
# Display first 10 rows of data.

data.head()

,name,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,...,fta,ft,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs
0,Brandon Ingram,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,...,2.3,69.9,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0
1,Andrew Harrison,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,...,3.4,76.5,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0
2,JaKarr Sampson,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,...,1.3,67.0,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0
3,Malik Sealy,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,...,1.3,68.9,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1
4,Matt Geiger,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,...,1.9,67.4,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1


In [14]:
# Display number of rows, number of columns.
rows, columns = data.shape
print(f'Columns: {columns} \nRows: {rows}\n' ) 

data.info()

Columns: 21 
Rows: 1340

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-n

The total number of rows is 1340 and it has 21 columns.

In [16]:
# Display all column names.

data.columns

Index(['name', 'gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p',
       'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov',
       'target_5yrs'],
      dtype='object')

The following table provides a description of the data in each column. 

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played per game|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

* Based on the preceding tables, the name column is categorical, and the rest of the columns are numerical.

### Check for missing values

In [19]:
# Display the number of missing values in each column.
# Check whether each value is missing.
#Aggregate the number of missing values per column.

data.isna().sum()

name           0
gp             0
min            0
pts            0
fgm            0
fga            0
fg             0
3p_made        0
3pa            0
3p             0
ftm            0
fta            0
ft             0
oreb           0
dreb           0
reb            0
ast            0
stl            0
blk            0
tov            0
target_5yrs    0
dtype: int64

All columns in this dataset have 0 missing values.

Checking for missing values is an important step in data exploration. Missing values are not particularly useful, so it's important to handle them by cleaning the data.

## **Statistical tests** 



In [35]:
# Display percentage (%) of values for each class (1, 0) represented in the target column of this dataset.

perc_5, perct_l = round(data['target_5yrs'].value_counts(normalize=True)*100,2)
print(f'Career duration of at least five years(%): {perc_5}% \nCareer duration of less than five years: {perct_l}%')

Career duration of at least five years(%): 62.01% 
Career duration of less than five years: 37.99%


* About 62% of the values in the target columm belong to class 1, and about 38% of the values belong to class 0. In other words, about 62% of players represented by this data have an NBA career duration of at least five years, and about 38% do not.


* The dataset is not perfectly balanced, but an exact 50-50 split is a rare occurance in datasets, and a 62-38 split is not too imbalanced. However, if the majority class made up 90% or more of the dataset, then that would be of concern, and it would be prudent to address that issue through techniques like upsampling and downsampling.

If there is a lot more representation of one class than another, then the model may be biased toward the majority class. When this happens, the predictions may be inaccurate. 

## **Results and evaluation** 


Now, perform feature engineering, with the goal of identifying and creating features that will serve as useful predictors for the target variable, `target_5yrs`. 

### Feature selection

The following table contains descriptions of the data in each column:

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

To prepare the dataset for analysis, I removed the name variable, as it does not provide useful statistical information. I also created a new feature, minutes_per_game, by dividing the total minutes played (min) by the number of games played (gp), in order to standardize playing time across players. From the available columns, I selected a subset of performance-related metrics that capture key aspects of a player's contribution on the court: fgm (field goals made), 3p_made (three-pointers made), ftm (free throws made), reb (rebounds), ast (assists), stl (steals), blk (blocks), and tov (turnovers). These variables offer a comprehensive view of a player's offensive and defensive impact.

In [47]:
# Select the columns to proceed with and save the DataFrame in new variable `selected_data`.
# Include the target column, `target_5yrs`.

selected_data = data[['gp', 'min', 'pts','fgm', '3p_made', 'ftm', 'reb','ast','stl','blk','tov','target_5yrs']]

# Display the first few rows.

selected_data.head()


,gp,min,pts,fgm,3p_made,ftm,reb,ast,stl,blk,tov,target_5yrs
0,36,27.4,7.4,2.6,0.5,1.6,4.1,1.9,0.4,0.4,1.3,0
1,35,26.9,7.2,2.0,0.7,2.6,2.4,3.7,1.1,0.5,1.6,0
2,74,15.3,5.2,2.0,0.4,0.9,2.2,1.0,0.5,0.3,1.0,0
3,58,11.6,5.7,2.3,0.1,0.9,1.9,0.8,0.6,0.1,1.0,1
4,48,11.5,4.5,1.6,0.0,1.3,2.5,0.3,0.3,0.4,0.8,1


### Feature transformation

An important aspect of feature transformation is feature encoding. If there are categorical columns that you would want to use as features, those columns should be transformed to be numerical. This technique is also known as feature encoding.

Many types of models are designed in a way that requires the data coming in to be numerical. So, transforming categorical features into numerical features is an important step.


In this particular dataset, name is the only categorical column and the other columns are numerical. Given that name is not selected as a feature, all of the features that are selected at this point are already numerical and do not require transformation.

### Feature extraction

Display the first few rows containing containing descriptions of the data for reference. The table is as follows:

<center>

|Column Name|Column Description|
|:---|:-------|
|`name`|Name of NBA player|
|`gp`|Number of games played|
|`min`|Number of minutes played per game|
|`pts`|Average number of points per game|
|`fgm`|Average number of field goals made per game|
|`fga`|Average number of field goal attempts per game|
|`fg`|Average percent of field goals made per game|
|`3p_made`|Average number of three-point field goals made per game|
|`3pa`|Average number of three-point field goal attempts per game|
|`3p`|Average percent of three-point field goals made per game|
|`ftm`|Average number of free throws made per game|
|`fta`|Average number of free throw attempts per game|
|`ft`|Average percent of free throws made per game|
|`oreb`|Average number of offensive rebounds per game|
|`dreb`|Average number of defensive rebounds per game|
|`reb`|Average number of rebounds per game|
|`ast`|Average number of assists per game|
|`stl`|Average number of steals per game|
|`blk`|Average number of blocks per game|
|`tov`|Average number of turnovers per game|
|`target_5yrs`|1 if career duration >= 5 yrs, 0 otherwise|

</center>

In [45]:
# Display the first few rows of `selected_data` for reference.
selected_data.head()

,gp,min,fgm,3p_made,ftm,reb,ast,stl,blk,tov,target_5yrs
0,36,27.4,2.6,0.5,1.6,4.1,1.9,0.4,0.4,1.3,0
1,35,26.9,2.0,0.7,2.6,2.4,3.7,1.1,0.5,1.6,0
2,74,15.3,2.0,0.4,0.9,2.2,1.0,0.5,0.3,1.0,0
3,58,11.6,2.3,0.1,0.9,1.9,0.8,0.6,0.1,1.0,1
4,48,11.5,1.6,0.0,1.3,2.5,0.3,0.3,0.4,0.8,1


 A new posible feature, minutes_per_game, by dividing the total minutes played (min) by the number of games played (gp), in order to standardize playing time across players.

In [48]:
# Extract two features that would help predict target_5yrs.
# Create a new variable named `extracted_data`.

extracted_data = selected_data.copy()

extracted_data['min_per_games'] = extracted_data['min']/extracted_data['gp']

extracted_data["total_points"] = extracted_data["gp"] * extracted_data["pts"]

extracted_data["points_per_minutes"] = extracted_data["total_points"]/ (extracted_data["gp"]* extracted_data["min"])



In [51]:
# Remove any columns from `extracted_data` that are no longer needed.
extracted_data = extracted_data.drop(['gp','min','pts'],axis=1)


# Display the first few rows of `extracted_data` to ensure that column drops took place.

extracted_data.head()


,fgm,3p_made,ftm,reb,ast,stl,blk,tov,target_5yrs,min_per_games,total_points,points_per_minutes
0,2.6,0.5,1.6,4.1,1.9,0.4,0.4,1.3,0,0.761111,266.4,0.270073
1,2.0,0.7,2.6,2.4,3.7,1.1,0.5,1.6,0,0.768571,252.0,0.267658
2,2.0,0.4,0.9,2.2,1.0,0.5,0.3,1.0,0,0.206757,384.8,0.339869
3,2.3,0.1,0.9,1.9,0.8,0.6,0.1,1.0,1,0.200000,330.6,0.491379
4,1.6,0.0,1.3,2.5,0.3,0.3,0.4,0.8,1,0.239583,216.0,0.391304


## Export data


In [56]:
# Export the extracted data.

extracted_data.to_csv('nba_data.csv', index = 0)

## **Considerations**


It is important to check for class balance in a dataset, particularly in the context of feature engineering and predictive modeling. If the target column in a dataset has more than 90% of its values belonging to one class, it is recommended to redistribute the data; otherwise, once a model is trained on the imbalanced data and predictions are made, the predictions may be biased.
Feature selection involves choosing features that help predict the target variable and removing columns that may not be helpful for prediction. In this process, and throughout feature engineering, it is important to make ethical considerations.
Feature transformation involves transforming features so that they are more usable for future modeling purposes, which includes encoding categorical features to turn them into numerical features.
Feature extraction involves combining existing columns meaningfully to construct new features that would help improve prediction.

The following attributes about player performance could help predict their NBA career duration and should be included in a presentation to stakeholders: field goals, three-point field goals, free throws, rebounds, assists, steals, blocks, turnovers, total points, and efficiency as points per minute.
It would be important to explain that these attributes, along with a relevant dataset, will be used in the next stage of the project. At that point, a model will be built to predict a player's career duration. Insights gained will be shared with stakeholders once the project is complete. Stakeholders would also appreciate being provided with a timeline and key deliverables that they can expect to receive.